### ISSUE TITLE
reading of seconds on Nordic Type 4 lines

### ISSUE THREAD

**Issue by WayneCrawford** [obspy/obspy#2348]
obspy incorrectly reads pick times (specifically seconds) when the seconds are given to millisecond precision (I assume this happens when I use high-accuracy mode, but in any case it is a "standard" output for SEISAN as can be seen in the definition of Nordic files (screenshot attached).

I have attached an example Nordic file: the first and third "events" are in this high-accuracy notation.  I have also attached a script to read this file and print out the pick times so that you can compare for yourself(ves).  I had to add the suffix "txt" so that it could be attached.

The solution is to read in line[22:28] as seconds instead of line[23:28], in obspy.io.nordic.core, I have attached a screenshot of the relevant code from the current git in which the relevant lines are 745 and 749

I am using obspy version 1.1.1, installed using conda (and conda-forge)

![nordic_type4line](https://user-images.githubusercontent.com/3601379/53949060-90981180-40c9-11e9-97b2-ec3f562365a0.png)
[correct_10_11.txt](https://github.com/obspy/obspy/files/2940653/correct_10_11.txt)
[nordic_example.py.txt](https://github.com/obspy/obspy/files/2940666/nordic_example.py.txt)
![nordic_code_example](https://user-images.githubusercontent.com/3601379/53949728-f5a03700-40ca-11e9-9dc7-64cf5c1de41f.png)

**Comment by megies** [comment]
@calum-chamberlain you did the nordic stuff, right? Can you have a look?

**Comment by calum-chamberlain** [comment]
@WayneCrawford Apologies and thanks for pointing that out - do you want to make a pull request to fix this? Happy to do it for you if you don't want to.  It would be good to add a high-precision test file as well (I don't have one, probably hence this popping up). It looks like [line 749](https://github.com/obspy/obspy/blob/master/obspy/io/nordic/core.py#L749) has a typo as well (should be `line[23:28]`, or rather `line[22:28]` as you state).

On that note @WayneCrawford if you are an extensive seisan user - do you have any other interesting files to add, or have noticed anything else not working as it should in Nordic? Any contributions on the nordic front would be really welcome!

**Comment by megies** [comment]
This sounds like a quick fix, so I'll add it to 1.2.0 which we will try to finalize on 15th, i.e. next Friday.

**Comment by WayneCrawford** [comment]
Thanks for doing this.  I saw the discrepancy on line 749 as well but, as I didn't know what specific case that handled, I wasn't sure it was a typo.  I'm not a very strong SEISAN user so I don't think I have any interesting files to contribute, but I will keep an eye out.  

The only other problem I was aware of with Nordic File reading was the existence of "seconds" >= 60, it seems your line 749 tries to handle this but I'm not sure it is fully complete or robust.  First, I've seen seconds > 100! (see attached example).  Then, it looks like pick_hour is being placed in the "minutes" slot.  I would replace with something like:

        seconds=float(line[22:28])
        if seconds<60. :
            time = UTCDateTime(evtime.year, evtime.month, evtime.day,
                               pick_hour, int(line[20:22]),
                               seconds) + day_add
        else:
             time = UTCDateTime(evtime.year, evtime.month, evtime.day,
                               pick_hour, int(line[20:22]),
                               seconds-60.) +60. + day_add

I don't even know if SEISAN limits itself to <120 seconds: if not, the second part would have to be a bit more clever...
 [nordic_crazyseconds.txt](https://github.com/obspy/obspy/files/2945603/nordic_crazyseconds.txt)
          
I'm still not very familiar with GitHub, pulls, and, above all, testing what I've done, so I'd prefer it if one of you could do so.  If not, I can give it a try!

**Comment by d-chambers** [comment]
> I don't even know if SEISAN limits itself to <120 seconds: if not, the second part would have to be a bit more clever...

We added a `strick` keyword to the `UTCDateTime` object for cases such as these in #2232. So as long as `strict` == `False` you can just pass as many seconds as you like to the `UTCDateTime` constructor like so:

```python
from obpsy import UTCDateTime

time = UTCDateTime(year=2001, month=1, day=1, hour=2, minute=23, second=14242, strict=False)

```
This will make the seisan code considerably simpler.

**Comment by calum-chamberlain** [comment]
@WayneCrawford all good, thanks for that info - I will make a PR this weekend.

Thanks for that pointer @d-chambers as well, that will make it much nicer! Wish I had spotted that.

**Comment by calum-chamberlain** [comment]
Implemented a fix in #2350 - I didn't end up using the strict argument, it was easier to add the seconds as a float after instantiating the UTCDateTime object.  The string formatting is really ugly in nordic, I would like to clean that up at some point...

@WayneCrawford could you let me know whether seisan is happy reading *high-accuracy* phase times when not run in high-accuracy mode?  I just want to know if it is okay to, by default, write out high-accuracy phase times?

**Comment by WayneCrawford** [comment]
Hi Calum,

I don’t really know what SEISAN is happy reading where: the docs are not clear and I could not find a bit of code that is always called to read in picks.  However, one should be able to count on the convention for Nordic type 4 lines specified in the SEISAN documents (see attached: the important part is that the seconds can be in columns 23-28 (slice 22:28)).  What seems crazy to me is that in the file nordic_crazyseconds.txt that I attached to one of the earlier posts, it looks like SEISAN is putting the seconds in columns 24:29 (for the case of seconds >=100 but not high resolution), which is definitely not recommended anywhere I see!

Regards
Wayne
_______________________________
Wayne Crawford
CNRS Researcher
Institut de Physique du Globe de Paris
1, rue Jussieu
75238 Paris Cedex 05
+33 6 5151 1054

On 10 March 2019 at 00:58:54, Calum Chamberlain (notifications@github.com) wrote:

Implemented a fix in #2350 - I didn't end up using the strict argument, it was easier to add the seconds as a float after instantiating the UTCDateTime object. The string formatting is really ugly in nordic, I would like to clean that up at some point...

@WayneCrawford could you let me know whether seisan is happy reading high-accuracy phase times when not run in high-accuracy mode? I just want to know if it is okay to, by default, write out high-accuracy phase times?

**Comment by calum-chamberlain** [comment]
Thanks for reminding me of that weirdness - if 29 is *supposed* to be free, but sometimes used by seconds we can happily read that.  Maybe that is an overflow from formatting at 5.2f in "standard" precision. I don't really get how SEISAN does time internally.  Because 29 is supposed to be free, I don't see the harm in trying to read from that column to cope with this overflow.
As an aside, one thing that makes this... interesting is the spec saying that seconds should be formatted as F6.0, which I don't think it ever is (it's either F6.3 or F5.2 depending on accuracy). I tried reading in high-accuracy style pick times without changing anything in SEISAN and it plotted and wrote at the correct time, so I'm happy defaulting to high-accuracy, but I will add a flag to run high-accuracy off.
I based on the wrong branch, so #2350 is closed now, and the work re-done in #2351.
